textproto/header.go

```diff
@@ -425,14 +425,34 @@ const (
 	maxLineOctets  = 4000
 )
 
+// ReadOptions represents the options in a header reader.
+type ReadOptions struct {
+	MaxHeaderLineLength int
```

Comment:
I'd prefer the limit to be per-header field (`MaxHeaderFieldLength`) instead of per-line, since this is what users will manipulate in the end.

Reply:
Does the name should be updated only, or implementation should be changed as well?

If implementation should be updated, I am not sure about the next step.

Reply:
Sorry, this comment somehow slipped through the cracks of my inbox.

Up until now, `maxLineOctets` was private and not exposed to users. Now that we're exposing the limits to users, we need to ask ourselves: what kind of limits should we expose?

- Maximum line length (current `maxLineOctets`)
- Maximum number of header lines (current `maxHeaderLines`)
- Maximum number of header fields
- Maximum header field length
- Maximum size of the full header

What will users care about?

IMHO users won't acre too much about the line length or the number of lines. Instead they may care about the whole header size and the number of header fields.

cc @foxcpp

Reply:
When contributing length limitations, I was concerned about the whole header size, that's right. Limiting header lines & line length in two places seemed to be the easiest way to implement this (effectively limiting it to maxLineOctets\*maxHeaderLines bytes).

Reply:
... I already mentioned it in 2 months ago :)

Reply:
We could use something like `io.LimitedReader` before passing the reader to `ReadHeader` to achieve the same result (maybe with a custom error when the limit is reached).

Reply:
Hm. Why do you always ruin my day (/s) by pointing out simple solutions to my problems? :)

I definitely could wrap io.Reader received from go-smtp with io.LimitReader. 
The only possible concern is redundant buffering if it becomes bufio.Reader(io.LimitedReader(bufio.Reader)).  (bufio.NewReader is no-op if argument is already bufio.Reader).

Reply:
Thanks guys for the comments. @foxcpp I just understood today what you mentioned two months ago :)

For the purposes of the issue I opened, I don't think adding reader options is a good option now.

If this sounds good to you guys, I think we can just return `TooBigError` when total length of the bytes read reaches the limit `(maxHeaderLines * maxLineOctets)`, Instead of checking header lines and line octets in two different functions.

I am waiting green light from you guys to update the PR in case my last suggestion is not a good option.

Reply:
Does it sound good to update the PR, according to the last comment?

Reply:
I'd prefer to have a single `MaxHeaderBytes` limit, just like `net/http` has.

Reply:
Sure, would it be okay to check if total length of bytes read equal to `MaxHeaderBytes` (4000000) instead of checking for `maxHeaderLines` and `maxLineOctets` separately ?

Reply:
Yeah. As I mentioned above, using an `io.LimitedReader` could be enough.